First refactoring work, decouple wireguard from clustering

kaiyou opened the pull request:
This first PR is still a WIP awaiting feedback.

The overall work is split into mostly consistent commits for easier incremental review, attempting to:
- remove the use of wireguard objects in cluster.go by handling metadata in a separate node.go and generating the overlay address in main.go
- stop passing the config object around and use function arguments directly
- lower the complexity of newCluster by moving pre-computations to separate functions

At this point, end to end tests run fine, I have not added any logic yet, and am waiting for the first refactoring work to be (hopefully) merged before actually adding features.

fixes #37

kaiyou commented:
Also, this is directly related to #37

costela commented on node.go:
```diff
@@ -0,0 +1,67 @@
+		}
+		node.nodeMeta = meta
+		nodes = append(nodes, node)
+		logrus.Infof("\taddr: %s, overlay: %s, pubkey: %s", node.Addr, node.OverlayAddr, node.PubKey)
```
I'm not a big fan of logging directly from "deep" functions. Can we maybe readd `multierr` and move the logging to main?

Ideally we'd do the same for `encodeNodeMeta`, but that one is trickier since it's wrapped by `getMeta`.

kaiyou replied:
Agreed. I quickly moved this logic to the node code, but it is not actually node specific logic, but more of a simple loop calling the metadata decoding. I suggest simply getting this loop back into the main loop, we will have to iterate in the main loop for logging anyway.

costela replied:
yeah, makes sense.

kaiyou commented on cluster.go:
```diff
@@ -194,41 +138,59 @@ func (c *cluster) members() (<-chan []node, <-chan error) {
+		if err != nil {
+			return nil, err
+		}
+		if isatty.IsTerminal(os.Stdout.Fd()) {
```
Of course, my bad for removing it.

kaiyou commented on cluster.go:
```diff
@@ -194,41 +138,59 @@ func (c *cluster) members() (<-chan []node, <-chan error) {
 
-func (n *node) String() string {
-	return n.Addr.String()
+func computeBindAddr(bindAddr string, bindIface string) (string, error) {
```
Agreed about moving this logic to the config code

kaiyou commented on main.go:
```diff
@@ -34,12 +35,21 @@ func main() {
 	}
 
-	cluster, err := newCluster(config, wg)
+	getMeta := func(limit int) []byte {
```
I disagree with implementing the interface. Pushing cluster-related concerns in the wireguard code and types is exactly what I would like to avoid. Implementing the interface in `wgState` would probably make the code a bit more concise, but it would complexify understanding, and then changes in `memberlist` will affect both the cluster *and* the wireguard code.

I intend (I have it working locally, just writing some more documentation before I commit) to move both wireguard and cluster code to separate module. It might be a bit more verbose, with additional `use` especially, but it is an assurance that we avoid mixing concerns too much: the glue code must be written in the main loop.

kaiyou replied:
I ended up making it an `Encode` function of the common `Node` type, which makes sense imo. And a `localNode` is passed to the cluster instance. The code is on a separate branch where I split the modules, so I'll push it later.

costela replied:
ok, that sounds reasonable.

Could `Node` then implement `memberlist.Delegate`? My problem is that I never liked having the cluster implement `Delegate` itself. It was a lazy "solution" on my part.

Do you think we could clear this up before going for the split PR? Or are the changes too intertwined?

kaiyou replied:
Definitely yes. `Node` being a shared type, `cluster.go` can have it implement the delegate interface, that is probably the cleanest way to do it. I can try and change it before splitting, I'll resolve any conflict later on :)

kaiyou commented:
I have a commit on the way for splitting things into modules. Should I wait for us to settle the current issues before pushing further bits into this PR maybe?

kaiyou commented:
Thanks again for the feedback. I just pushed most requested changes, hope this is still in the right direction. Waiting for your input before I push anything about splitting into modules.

costela commented on config.go:
```diff
@@ -44,10 +44,26 @@ func loadConfig() (*config, error) {
+		// Compute the actual bind address based on the provided interface
+		iface, err := net.InterfaceByName(config.BindIface)
+		if err != nil {
+			return nil, fmt.Errorf("cannot find interface %s", config.BindIface)
```
this is an improvement on the original version, but I think we should still wrap the original error and not assume what it's about:
```suggestion
			return nil, fmt.Errorf("could not get interface by name %s: %w", config.BindIface, err)
```

costela commented on config.go:
```diff
@@ -44,10 +44,26 @@ func loadConfig() (*config, error) {
+		}
+		addrs, err := iface.Addrs()
+		if err != nil {
+			return nil, fmt.Errorf("no available ip address on interface %s", config.BindIface)
```
same as above:
```suggestion
			return nil, fmt.Errorf("could not get addresses for interface %s: %w", config.BindIface, err)
```

kaiyou commented:
Thanks for the feedback, and no problem about the delays.

After a second thought, passing a `Node` instance to the cluster as the local node is implemented on top of many changes for splitting the code into packages, and I fear the number of conflicts that will arise. Would you mind if I updated the PR with these changes?

costela commented:
>  Would you mind if I updated the PR with these changes?

Not at all! Could just mean I'll need a bit more time to review them, but that's not a deal-breaker.

kaiyou commented:
Just pushed the remaining changes. I hope this is not too much for review.

I tried implementing the delegate interface on the `Node` type by subtyping in the cluster package. I honestly am not sure this improves readability or maintainability. It does make more sense I agree. Up to you I guess :)

costela commented on main.go:
```diff
@@ -29,19 +34,32 @@ func main() {
+	localNode := common.Node{}
+	localNode.OverlayAddr = wgstate.OverlayAddr
+	localNode.PubKey = wgstate.PubKey.String()
+	cluster.Update(localNode)
```
Something about this block feels off :thinking: 

How about this: unexport `AssignOverlayAddr()` and call it from `wg.New`, expanding it to also require the `cluster.LocalName` and `OverlayNet` (we can swap the instantiation order of `wgstate` and `cluster`, now that the latter doesn't require the former anymore).
Than we can add `wgstate.LocalNode()` that returns a `Node` with the appropriate info, and finally rename `cluster.Update` to something a bit more descriptive (maybe `Announce`? or `SetLocalNode`?)

kaiyou replied:
I would prefer a `wgstate.UpdateNode()` and creating the node in the main loop itself. It will allow for something else than the wireguard code to update the code, say the local routing table providing additional routes, as already asked in issue #28 .

Agreed otherwise, I will implement this right aways.

kaiyou commented on common/node.go:
```diff
@@ -0,0 +1,51 @@
+package common
```
I am undecided on the matter, so let's go for a `node` package and `node.Info` type :)

kaiyou replied:
After some tests, I really would recommend to stick with `common`, or maybe find another name for that module. Naming it `node` conflicts with local variables named `node`, and renaming those lowers readability.

I would personally stick with `common` and admin that this is not a very reusable module, that we will just use to hold types used by other modules, not really intended to be reused outside wesher directly.

costela replied:
ok, I can live with that

costela commented on main.go:
```diff
@@ -29,19 +34,31 @@ func main() {
 
-	nodec, errc := cluster.members() // avoid deadlocks by starting before join
+	// Assign a local node address and propagate it to the cluster
+	localNode := &common.Node{}
+	wgstate.UpdateNode(localNode)
+	cluster.SetLocalNode(localNode)
+	cluster.Update()
```
I overlooked one important detail and asked you to remove `LocalName` from `wg.State`. In hindsight that was not a good suggestion: now these 3 lines are pretty densely packed: "update, set, update", where the first two actually modify `localNode` in non-obvious ways. Doesn't feel as clean as the rest of the refactoring. My bad.

How about the following:
- return the local node from `wg.New` for now and deal with the requirements of #28 when we get to it
- reinstante `wg.State.localName`, so we can get rid of the awkward name setting in `SetLocalNode`
- merge `SetLocalNode` and `Update` (the call to `SetLocalNode` in the constructor can then use the reinstanted `State.localName` directly).

This way the end-product here would be the considerably more readable:
```suggestion
	cluster.SetLocalNode(localNode)
```

kaiyou commented:
Thank you again for the review. I have tried and implemented your suggestions. I will wait for the build. The state test will possibly fail depending on permissions on the builder.

costela commented on cluster/state_test.go:
```diff
@@ -0,0 +1,32 @@
+		},
+	}
+
+	cluster.state.save()
```
```suggestion
	if err := cluster.state.save(); err != nil {
		t.Error(err)
	}
```

If the error is really just a permission issue on the folder, we could just make `statePath` a variable and change it to `/tmp/wesher.state.json` or something like that during testing.

This will change after #38 anyway.

kaiyou commented:
Pushed these last details and fixed the tests. Also, I removed the WIP from the title since I feel confident about this. Thanks for guiding the work :)

costela commented:
Nice! This is a definite improvement and will make further work easier!

Thanks again! :+1:
